File: tests/benchmarks/coro/producer_consumer/producer_consumer.erl
%% Producer-Consumer benchmark - Erlang equivalent
%% M producers send to a shared process, N consumers receive
%% Uses a broker process to simulate buffered channel
%% Usage: erl -noshell -s producer_consumer main -s init stop
-module(producer_consumer).
-export([main/0, main/3]).

main() ->
    main(10, 10, 100000).

main(Producers, Consumers, Items) ->
    io:format("=== 生产者消费者测试 (Erlang) ===~n"),
    io:format("生产者: ~p~n", [Producers]),
    io:format("消费者: ~p~n", [Consumers]),
    io:format("总数量: ~p~n", [Items]),

    Self = self(),
    T0 = erlang:monotonic_time(millisecond),

    %% Start broker (simulates buffered channel)
    Broker = spawn(fun() -> broker_loop(queue:new(), queue:new(), 0) end),

    ItemsPerProducer = Items div Producers,

    %% Start producers
    ProducerPids = [spawn(fun() ->
        producer_loop(Id, ItemsPerProducer, Broker),
        Self ! {producer_done, Id}
    end) || Id <- lists:seq(0, Producers - 1)],

    %% Start consumers
    ConsumerPids = [spawn(fun() ->
        Count = consumer_loop(Broker, 0),
        Self ! {consumer_done, Id, Count}
    end) || Id <- lists:seq(0, Consumers - 1)],

    %% Wait for all producers
    wait_producers(Producers),

    %% Signal broker to close
    Broker ! close,

    %% Collect consumer results
    TotalConsumed = collect_consumers(Consumers, 0),

    T1 = erlang:monotonic_time(millisecond),
    Elapsed = T1 - T0,
    io:format("消费总数: ~p~n", [TotalConsumed]),
    io:format("时间: ~p ms~n", [Elapsed]),
    case Elapsed > 0 of
        true ->
            io:format("吞吐量: ~p items/sec~n", [TotalConsumed * 1000 div Elapsed]);
        false ->
            io:format("吞吐量: 太快无法测量~n")
    end.

%% Broker: buffers items, dispatches to waiting consumers
%% Uses queue for both items and waiters (O(1) amortized)
broker_loop(Queue, WaitQ, Closed) ->
    receive
        {send, Item} ->
            case queue:out(WaitQ) of
                {{value, Waiter}, WaitQ2} ->
                    Waiter ! {item, Item},
                    broker_loop(Queue, WaitQ2, Closed);
                {empty, _} ->
                    broker_loop(queue:in(Item, Queue), WaitQ, Closed)
            end;
        {recv, Consumer} ->
            case queue:out(Queue) of
                {{value, Item}, Queue2} ->
                    Consumer ! {item, Item},
                    broker_loop(Queue2, WaitQ, Closed);
                {empty, _} when Closed =:= 1 ->
                    Consumer ! closed,
                    broker_loop(Queue, WaitQ, Closed);
                {empty, _} ->
                    broker_loop(Queue, queue:in(Consumer, WaitQ), Closed)
            end;
        close ->
            drain_waiters(WaitQ),
            broker_loop(Queue, queue:new(), 1)
    end.

drain_waiters(WaitQ) ->
    case queue:out(WaitQ) of
        {{value, W}, Rest} -> W ! closed, drain_waiters(Rest);
        {empty, _} -> ok
    end.

producer_loop(_Id, 0, _Broker) ->
    ok;
producer_loop(Id, Count, Broker) ->
    Broker ! {send, Id * 1000000 + Count},
    producer_loop(Id, Count - 1, Broker).

consumer_loop(Broker, Count) ->
    Broker ! {recv, self()},
    receive
        {item, _Item} ->
            %% Simulate processing
            consumer_loop(Broker, Count + 1);
        closed ->
            Count
    end.

wait_producers(0) -> ok;
wait_producers(N) ->
    receive {producer_done, _} -> wait_producers(N - 1) end.

collect_consumers(0, Total) -> Total;
collect_consumers(N, Total) ->
    receive {consumer_done, _, Count} -> collect_consumers(N - 1, Total + Count) end.

File: tests/benchmarks/coro/producer_consumer/producer_consumer.go
// 作者：xingleixu@gmail.com
// 生产者消费者测试
// M 个生产者生产数据，N 个消费者消费数据

package main

import (
	"fmt"
	"os"
	"strconv"
	"sync"
	"sync/atomic"
	"time"
)

func main() {
	PRODUCERS := 10 // 生产者数量
	CONSUMERS := 10 // 消费者数量
	ITEMS := 100000 // 总生产数量

	if len(os.Args) > 1 {
		if n, err := strconv.Atoi(os.Args[1]); err == nil {
			PRODUCERS = n
		}
	}
	if len(os.Args) > 2 {
		if n, err := strconv.Atoi(os.Args[2]); err == nil {
			CONSUMERS = n
		}
	}
	if len(os.Args) > 3 {
		if n, err := strconv.Atoi(os.Args[3]); err == nil {
			ITEMS = n
		}
	}

	fmt.Println("=== 生产者消费者测试 ===")
	fmt.Println("生产者:", PRODUCERS)
	fmt.Println("消费者:", CONSUMERS)
	fmt.Println("总数量:", ITEMS)

	queue := make(chan int, 100)
	var totalConsumed int64

	var producerWg sync.WaitGroup
	var consumerWg sync.WaitGroup

	itemsPerProducer := ITEMS / PRODUCERS
	remainingItems := ITEMS % PRODUCERS

	start := time.Now()

	// 启动生产者
	for i := 0; i < PRODUCERS; i++ {
		producerWg.Add(1)
		go func(id int) {
			defer producerWg.Done()
			count := itemsPerProducer
			if id < remainingItems {
				count++
			}
			for j := 0; j < count; j++ {
				queue <- id*1000000 + j
			}
		}(i)
	}

	// 等待生产者完成后关闭队列
	go func() {
		producerWg.Wait()
		close(queue)
	}()

	// 启动消费者
	for i := 0; i < CONSUMERS; i++ {
		consumerWg.Add(1)
		go func(id int) {
			defer consumerWg.Done()
			var consumed int64
			for item := range queue {
				// 模拟处理
				_ = item * 2
				consumed++
			}
			atomic.AddInt64(&totalConsumed, consumed)
		}(i)
	}

	// 等待消费者完成
	consumerWg.Wait()

	elapsed := time.Since(start)
	elapsedMs := float64(elapsed) / float64(time.Millisecond)

	fmt.Println("消费总数:", totalConsumed)
	fmt.Printf("时间: %.3f ms\n", elapsedMs)
	if elapsed > 0 {
		fmt.Println("吞吐量:", int(float64(totalConsumed)/elapsed.Seconds()), "items/sec")
	} else {
		fmt.Println("吞吐量: 太快无法测量")
	}
}
